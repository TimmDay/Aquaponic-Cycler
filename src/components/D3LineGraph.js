import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import * as d3 from 'd3';
import { select } from 'd3-selection';
import tip from 'd3-tip';
import { startSetCycleData, updateSelectedNodeData } from '../actions/aquaponic';
import { handleToggleModalEditCycle } from './../actions/ux';

class D3LineGraph extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      graphWidth:
        this.props.windowWidth -
        this.props.margin.left -
        this.props.margin.right,
      graphHeight:
        this.props.windowHeight -
        this.props.margin.top -
        this.props.margin.bottom,
      colors: ['#4286f4', '#2ba85f', '#a564e5', '#f23551'],
      data: ''
    };
  }
  // realtime firestore database
  listenToData = () => this.props.startSetCycleData(); 

  componentDidMount() {
    this.createGraph();
  }

  componentDidUpdate() {
    // filter data into the 4 data streams, sort by date
    const dataArr = [
      this.props.cyclingData
        .filter(d => d.name === 'pH')
        .sort((a, b) => new Date(a.date) - new Date(b.date)),
      this.props.cyclingData
        .filter(d => d.name === 'ammonia')
        .sort((a, b) => new Date(a.date) - new Date(b.date)),
      this.props.cyclingData
        .filter(d => d.name === 'nitrite')
        .sort((a, b) => new Date(a.date) - new Date(b.date)),
      this.props.cyclingData
        .filter(d => d.name === 'nitrate')
        .sort((a, b) => new Date(a.date) - new Date(b.date))
    ];

    // use a separate update function depending on data selection
    switch (this.props.selectedFilter) {
      case 'pH':
        this.update(dataArr[0], 'pH');
        break;
      case 'ammonia':
        this.update(dataArr[1], 'ammonia');
        break;
      case 'nitrite':
        this.update(dataArr[2], 'nitrite');
        break;
      case 'nitrate':
        this.update(dataArr[3], 'nitrate');
        break;
      case 'view all':
        this.update(dataArr[0], 'pH');
        this.update(dataArr[1], 'ammonia');
        this.update(dataArr[2], 'nitrite');
        this.update(dataArr[3], 'nitrate');
        select('.aquaponic__y-axis').style('opacity', 0); // hide y axis for view all
        break;
      default:
        break;
    }
  }

  createGraph = () => {
    const svg = select('.line-graph')
      .append('svg')
      .attr('width', this.props.windowWidth)
      .attr('height', this.props.windowHeight);

    const graph = svg
      .append('g')
      .attr('width', this.state.graphWidth)
      .attr('height', this.state.graphHeight)
      .attr(
        'transform',
        `translate(${this.props.margin.left}, ${this.props.margin.top})`
      )
      .attr('class', 'graph');

    const xAxisGroup = graph
      .append('g')
      .attr('class', 'aquaponic__x-axis')
      .attr('transform', `translate(0,${this.state.graphHeight})`); //move from top (default) to bottom

    const yAxisGroup = graph.append('g').attr('class', 'aquaponic__y-axis');

    const path1 = graph.append('path').attr('class', 'aquaponic__line1');
    const path2 = graph.append('path').attr('class', 'aquaponic__line2');
    const path3 = graph.append('path').attr('class', 'aquaponic__line3');
    const path4 = graph.append('path').attr('class', 'aquaponic__line4');

    const dottedLineGroup = graph
      .append('g')
      .attr('class', 'aquaponic__dotted')
      .style('opacity', 0);
    dottedLineGroup
      .append('line')
      .attr('class', 'aquaponic__dotted-x')
      .attr('stroke', '#bbb')
      .attr('stroke-width', 2)
      .attr('stroke-dasharray', 4);
    dottedLineGroup
      .append('line')
      .attr('class', 'aquaponic__dotted-y')
      .attr('stroke', '#bbb')
      .attr('stroke-width', 2)
      .attr('stroke-dasharray', 4);

    this.listenToData(); 
  };

  update(data, param) {
    const graph = select('.graph');
    const xAxisGroup = select('.aquaponic__x-axis');
    const yAxisGroup = select('.aquaponic__y-axis');
    const path1 = select('.aquaponic__line1');
    const path2 = select('.aquaponic__line2');
    const path3 = select('.aquaponic__line3');
    const path4 = select('.aquaponic__line4');
    const dottedLines = select('.aquaponic__dotted');
    const dottedX = select('.aquaponic__dotted-x');
    const dottedY = select('.aquaponic__dotted-y');
    select('.aquaponic__tip').remove(); //remove old tip element

    const toolTip = tip()
      .attr('class', 'aquaponic__tip')
      .html(d => {
        let content = `<div class='aquaponic__tip-details'>
        ${d.value} ${this.props.selectedFilter === 'pH' ? '' : 'ppm'}
        </div>`;
        content += `<div class='aquaponic__tip-details'>
        ${moment(new Date(d.date)).format('ddd, hA')}
        </div>`;
        return content;
      });
    graph.call(toolTip); //apply new tip el to the graph element

    // scales
    const x = d3.scaleTime([0, this.state.graphWidth]); //timescale
    const y = d3.scaleLinear([this.state.graphHeight, 0]);
    const line = d3
      .line()
      .x(d => x(new Date(d.date)))
      .y(d => y(d.value))
      .curve(d3.curveMonotoneX);
      //.curve(d3.curveCardinal.tension(0.7));
      // .curve(d3.curveCatmullRom.alpha(0.9));

    // 1. scale domains. set up axes.

    // set scale domains
    x.domain(d3.extent(this.props.dateExtent, d => new Date(d))); 
    // x.domain(d3.extent(data, d => new Date(d.date))); //extent returns [min, max]. Dates stored as strings, so bring them back to timestamps for comparison
    y.domain([-0.1, d3.max(data, d => parseFloat(d.value)) + 0.2]);

    //update circles and lines, for the 1 of 4 lines that is selected
    if (param === 'pH') {
      path1
        .data([data])
        .attr('fill', 'none')
        .attr('stroke', this.state.colors[0])
        .attr('stroke-width', 2)
        .attr('d', line)
        .style('opacity', 1);

      const circles = graph.selectAll('.circle-ph').data(data); // 2. join data
      circles.exit().remove(); // 3. exit selection

      // 4. update existing dom els
      circles
        .attr('cx', d => x(new Date(d.date)))
        .attr('cy', d => y(d.value))
        .attr('fill', () => this.state.colors[0])
        .style('opacity', 1);

      // 5. enter selection
      circles
        .enter()
        .append('circle')
        .attr('class', 'circle-ph')
        .attr('r', 4)
        .attr('cx', d => x(new Date(d.date))) // x axis is dates. we made x func (timeScale) to take a date obj within the defined xaxis limits.
        //date stored as string, so coerce it to date type with new Date
        .attr('cy', d => y(d.value)) // y axis is values. have a ya scale func also, that takes a value (linearScale)
        .attr('fill', () => this.state.colors[0]);

    } else if (param === 'ammonia') {
      path2
        .data([data])
        .attr('fill', 'none')
        .attr('stroke', this.state.colors[1])
        .attr('stroke-width', 2)
        .attr('d', line)
        .style('opacity', 1);

      const circles = graph.selectAll('.circle-amm').data(data); // 2. join data
      circles.exit().remove(); // 3. exit selection

      // 4. update existing dom els
      circles
        .attr('cx', d => x(new Date(d.date)))
        .attr('cy', d => y(d.value))
        .attr('fill', () => this.state.colors[1])
        .style('opacity', 1);

      // 5. enter selection
      circles
        .enter()
        .append('circle')
        .attr('class', 'circle-amm')
        .attr('r', 4)
        .attr('cx', d => x(new Date(d.date))) // x axis is dates. we made x func (timeScale) to take a date obj within the defined xaxis limits.
        //date stored as string, so coerce it to date type with new Date
        .attr('cy', d => y(d.value)) // y axis is values. have a ya scale func also, that takes a value (linearScale)
        .attr('fill', () => this.state.colors[1]);

    } else if (param === 'nitrite') {
      path3
        .data([data])
        .attr('fill', 'none')
        .attr('stroke', this.state.colors[2])
        .attr('stroke-width', 2)
        .attr('d', line)
        .style('opacity', 1);

      const circles = graph.selectAll('.circle-ite').data(data); // 2. join data
      circles.exit().remove(); // 3. exit selection

      // 4. update existing dom els
      circles
        .attr('cx', d => x(new Date(d.date)))
        .attr('cy', d => y(d.value))
        .attr('fill', () => this.state.colors[2])
        .style('opacity', 1);

      // 5. enter selection
      circles
        .enter()
        .append('circle')
        .attr('class', 'circle-ite')
        .attr('r', 4)
        .attr('cx', d => x(new Date(d.date))) // x axis is dates. we made x func (timeScale) to take a date obj within the defined xaxis limits.
        //date stored as string, so coerce it to date type with new Date
        .attr('cy', d => y(d.value)) // y axis is values. have a ya scale func also, that takes a value (linearScale)
        .attr('fill', () => this.state.colors[2]);

    } else if (param === 'nitrate') {
      path4
        .data([data])
        .attr('fill', 'none')
        .attr('stroke', this.state.colors[3])
        .attr('stroke-width', 2)
        .attr('d', line)
        .style('opacity', 1);

      const circles = graph.selectAll('.circle-ate').data(data); // 2. join data
      circles.exit().remove(); // 3. exit selection

      // 4. update existing dom els
      circles
        .attr('cx', d => x(new Date(d.date)))
        .attr('cy', d => y(d.value))
        .attr('fill', () => this.state.colors[3])
        .style('opacity', 1);

      // 5. enter selection
      circles
        .enter()
        .append('circle')
        .attr('class', 'circle-ate')
        .attr('r', 4)
        .attr('cx', d => x(new Date(d.date))) //date stored as string, so coerce it to date type with new Date
        .attr('cy', d => y(d.value)) // y axis is values. have a ya scale func also, that takes a value (linearScale)
        .attr('fill', () => this.state.colors[3]);
    }

    // create and call x-axis
    const xAxis = d3
      .axisBottom(x)
      .ticks(10)
      .tickFormat(d3.timeFormat('%b %d'));
    xAxisGroup.call(xAxis);
    xAxisGroup
      .selectAll('text')
      .attr('transform', 'rotate(-40)') // rotate axis text
      .attr('text-anchor', 'end');

    const yAxis = d3
      .axisLeft(y)
      .ticks(7)
      .tickFormat(
        d => `${d} ${this.props.selectedFilter === 'pH' ? '' : 'ppm'}`
      );

    yAxisGroup.style('opacity', 1); //re-display it, if it was hidden for view all
    yAxisGroup.call(yAxis);
    yAxisGroup.selectAll('text').attr('text-anchor', 'end');

    // USER INTERACTION - MOUSE
    graph
      .selectAll('circle')
      .on('mouseover', (d, i, n) => {
        toolTip.show(d, n[i]);
        select(n[i])
          .transition()
          .duration(100)
          .attr('r', 8);
        dottedX
          .attr('x1', 0)
          .attr('y1', y(d.value))
          .attr('x2', x(new Date(d.date)))
          .attr('y2', y(d.value));
        dottedY
          .attr('x1', x(new Date(d.date)))
          .attr('y1', this.state.graphHeight)
          .attr('x2', x(new Date(d.date)))
          .attr('y2', y(d.value));
          if (this.props.selectedFilter !== 'view all') dottedLines.style('opacity', 1);
      })
      .on('mouseleave', (d, i, n) => {
        toolTip.hide();
        select(n[i])
          .transition()
          .duration(300)
          .attr('r', 4);
        dottedLines.style('opacity', 0);
      })
      .on('click', (d) => {
        const nodeData = { id: d.id, date: d.date, name: d.name, value: d.value };
        this.props.updateSelectedNodeData(nodeData);
        this.props.handleToggleModalEditCycle();
      })
  }

  render() {
    return <div className="line-graph" />;
  }
}

const mapStateToProps = state => {
  return {
    selectedFilter: state.aquaponicReducer.selectedFilter,
    cyclingData: state.aquaponicReducer.cyclingData,
    dateExtent: state.aquaponicReducer.dateExtent
  };
};

const mapDispatchToProps = dispatch => ({
  startSetCycleData: () => dispatch(startSetCycleData()),
  handleToggleModalEditCycle: () => dispatch(handleToggleModalEditCycle()),
  updateSelectedNodeData: (obj) => dispatch(updateSelectedNodeData(obj))
});

export default connect(mapStateToProps, mapDispatchToProps)(D3LineGraph);

D3LineGraph.defaultProps = {
  margin: { top: 10, right: 70, bottom: 50, left: 70 },
  windowWidth: 600,
  windowHeight: 330
};