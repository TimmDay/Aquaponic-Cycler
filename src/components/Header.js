import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard">
          <h1>Aquaponic Cycler</h1>
          <p className="header_tagline">Tracking your bacterial food producers</p>
        </Link>

        <div className='header__link--container'>
          <Link className="header__link" to="/aquaponic">cycler</Link>
          <a className="header__link" href='#'>CV</a>
          <a className="header__link" href='www.timmday.com'>timmday.com</a>
          <a className="header__link" href='#'>github</a>
        </div>

        <button className="header__link" onClick={startLogout}>Logout</button> //TODO: RESPONSIVE
      </div>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);