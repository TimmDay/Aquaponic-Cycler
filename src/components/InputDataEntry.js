import React from 'react';
import { connect } from 'react-redux';
import {  startAddEntry } from '../actions/aquaponic';

class InputDataEntry extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      input: '',
      colors: ['#4286f4', '#2ba85f', '#a564e5', '#f23551'],
      error: ''
    };
  }

  chooseColor = (filter) => {
    switch (filter) {
      case 'pH':
        return this.state.colors[0];
      case 'ammonia':
        return this.state.colors[1];
      case 'nitrite':
        return this.state.colors[2];
      case 'nitrate':
        return this.state.colors[3];
      default:
        return;
    }
  }

  handleOnInputChange = evt => {
    evt.preventDefault();
    const value = evt.target.value;
    
    if (this.props.selected === 'pH') { 
      const regex = /^(1[0-3](\.\d{0,1})?)$|^\d?(\.\d{0,1})?$|^14$/; // 1 -> 14 with opt .0
      if (!value || value.match(regex)) {
          console.log(`ph: ${value}`);
          this.setState(() => ({ input: value, error: '' }));
        } else {
          this.setState(() => ({ error: `${this.props.selected} range is 1 -> 14` }));
        }
        
    } else if (this.props.selected === 'nitrate') { //0 -> 200 with opt .0
      const regex = /^1?\d?\d?(\.\d{0,1})?$|^200$/;
      if (!value || value.match(regex)) {
        console.log(`nitrate: ${value}`);
        this.setState(() => ({ input: value, error: '' }));
      } else {
        this.setState(() => ({ error: 'nitrate should be in the range 0 -> 120 ppm' }));
      }

    } else {
      const regex = /^10(\.0?)?$|^\d(\.\d{0,2})?$/; //ammonia or nitrite, 0 -> 10, with opt .00
      if (!value || value.match(regex)) {
        console.log(`amm/nitrite: ${value}`);
        this.setState(() => ({ input: value, error: '' }));
      } else {
        this.setState(() => ({ error: `${this.props.selected} should be in the range 0 -> 10 ppm, with up to 2 decimal places` }));
      }
    }
  }
  
  onSubmit = evt => {
    evt.preventDefault();
    const input = document.querySelector('.aquaponic__input');
    let val = input.value;
    if (val.endsWith('.')) val = val += '0'; // if entry ends in decimal, concat a 0
    const obj = {value: val, name: this.props.selected, date: new Date().toString()};
    this.props.startAddEntry(obj)
    this.setState(() => ({ input: '', error: '' }));
  };

  render() {
    return (
      <div className="content-container flex-container">
        <form
          className="aquaponic__form"
          onSubmit={this.onSubmit}
          autoComplete="off"
        >
          {this.props.selected !== 'view all' && (
            <p>
              {' '}
              How much{' '}
              <span style={{ color: this.chooseColor(this.props.selected) }}>
                {this.props.selected}
              </span>{' '}
              today?
            </p>
          )}

          {this.props.selected !== 'view all' && (
            <input
              autoFocus
              className="aquaponic__input"
              id="pH"
              onChange={this.handleOnInputChange}
              style={{
                borderBottom:
                  '1px solid ' + this.chooseColor(this.props.selected)
              }}
              type="text"
              value={this.state.input}
            />
          )}
          {this.state.error && (
            <p className="aquaponic__error-msg">{this.state.error}</p>
          )}
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddEntry: entry => dispatch(startAddEntry(entry)),
});

export default connect(undefined, mapDispatchToProps)(InputDataEntry);