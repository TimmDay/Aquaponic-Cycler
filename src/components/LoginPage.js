import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startLoginDemo } from '../actions/auth';

export const LoginPage = ({ startLogin, startLoginDemo }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Aquaponic Cycling</h1>
      <p>Stabilising Food Producing Ecosystems</p>
      <div className='login__btn-bar'>
        <button className="login__btn" onClick={startLogin}>Login with Google</button>
        <button className="login__btn" onClick={startLoginDemo}>View DEMO</button>
      </div>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
  startLoginDemo: () => dispatch(startLoginDemo('demo'))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);