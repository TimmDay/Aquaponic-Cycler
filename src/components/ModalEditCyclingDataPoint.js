import React from 'react';
import { connect } from 'react-redux';
import { SingleDatePicker } from 'react-dates'; // requires moment
import moment from 'moment';
import Modal from 'react-modal';
import { handleToggleModalEditCycle } from './../actions/ux';
import { startEditEntry } from './../actions/aquaponic';

class ModalEditCyclingDataPoint extends React.Component {

  constructor(props) {
      super(props);
      this.state = {
        name: '', 
        date: '', 
        value: ''
      }
  }

  componentDidUpdate(prevProps) {
    if (this.props.isModalOpenEditCycle && !prevProps.isModalOpenEditCycle) {      
      this.handlePreFillModalInputsOnOpen();
    }
  }
  handlePreFillModalInputsOnOpen = () => {
    this.setState(() => ({
      name: this.props.selectedNode.name === null ? '' : this.props.selectedNode.name, 
      date: this.props.selectedNode.date === null ? '' : this.props.selectedNode.date, 
      value: this.props.selectedNode.value === null ? '' : this.props.selectedNode.value,
      error: ''
    }));
  };

  handleOnClickReset = () => {
    this.setState({ name: '', date: '', value: '' });
    this.handlePreFillModalInputsOnOpen();
  }

  handleOnRequestClose = () => {
      this.handleOnClickReset();
      this.props.handleToggleModalEditCycle();
  };

  handleOnNameChange = (evt) => {
    evt.preventDefault();
    const name = evt.target.value;
    this.setState(() => ({name: name}))
};

  handleOnValueChange = (evt) => {
      evt.preventDefault();
      const value = evt.target.value;
      this.setState(() => ({value: value}))
  };

  onSubmitChanges = () => {
    const obj = {
      id: this.props.selectedNode.id,
      name: this.state.name,
      value: this.state.value,
      date: (typeof this.state.date === 'string') ? this.state.date : new Date(this.state.date).toString() //make sure that string
    };
    this.props.startEditEntry(obj);
    this.props.handleToggleModalEditCycle();
  }

  stampToMoment = (dateString) => {
    if (dateString) {
      console.log(new Date(dateString).getTime());
      const timeStampFromString = new Date(dateString).getTime();
      console.log(moment(timeStampFromString));
      
      return moment(timeStampFromString);
    }
  }
  momentToStamp = (mom) => {
    if (mom) {
      console.log(mom.valueOf());
      const newTimeStamp = mom.valueOf();
      return new Date(newTimeStamp)
    }
  }
    
  
  onDateChange = (pickerGenMoment) => {
    if (pickerGenMoment) { //prevents user clearing the value
        // because if date is cleared, this func gets called with nothing
        // (we don't want to update this part of state to nothing)
        const dateString = this.momentToStamp(pickerGenMoment);
        console.log(pickerGenMoment);
        console.log(dateString);

        this.setState(() => ({ date: dateString }));
    }
};
  onFocusChange = ({ focused }) => {
    this.setState(()=> ({ calendarFocused: focused }));
  };

  render () {
    return (
      <Modal
          isOpen= {this.props.isModalOpenEditCycle}
          contentLabel="Selected Option"
          onRequestClose={this.props.handleToggleModalEditCycle}
          closeTimeoutMS={200}
          className="modal"
          ariaHideApp={false}
      >
          <form
            className="modal__form"
            onSubmit={this.onSubmit}
          >
            {this.state.error && <p className="form__error">{this.state.error}</p>}

            <select
            className="modal__styled-select"
              value={this.state.name}
              onChange={this.handleOnNameChange}
            >
              <option>pH</option>
              <option>ammonia</option>
              <option>nitrite</option>
              <option>nitrate</option>
            </select>

            <SingleDatePicker
              date={this.stampToMoment(this.state.date)}
              onDateChange={this.onDateChange}
              focused={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false} //every day available, past and future
            />

            
            <input
              className="aquaponic__input"
              type="text"
              placeholder={this.props.selectedNode.value}
              value={this.state.value}
              onChange={this.handleOnValueChange}
            />
        </form>

        <div className="form__btn-bar">
            <button
                className="form__btn"
                onClick={this.handleOnRequestClose}
            >close
            </button>

            <button
                className="form__btn"
                onClick={this.handleOnClickReset}
            >reset
            </button>

            <button
                className="form__btn"
                onClick={this.onSubmitChanges}
            >submit changes
            </button>
        </div>
      </Modal>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    isModalOpenEditCycle: state.uxReducer.isModalOpenEditCycle,
    selectedNode: state.aquaponicReducer.selectedNode
  }
};

const mapDispatchToProps = (dispatch) => ({
  handleToggleModalEditCycle: () => dispatch(handleToggleModalEditCycle()),
  startEditEntry: (obj) => dispatch(startEditEntry(obj))
});

export default connect(mapStateToProps, mapDispatchToProps)(ModalEditCyclingDataPoint);