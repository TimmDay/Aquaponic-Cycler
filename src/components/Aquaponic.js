import React from 'react';
import { connect } from 'react-redux';
import { setSelectedFilter, startAddEntry } from '../actions/aquaponic';
import D3LineGraph from './D3LineGraph';

class Aquaponic extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: 'view all',
      input: '',
      error: ''
    };
  }

  handleOnClick = evt => {
    // update text of form span
    this.setState({ selected: evt.target.innerHTML });

    // remove active btn class from others
    const els = document.querySelectorAll('.aquaponic__btn');
    els.forEach(btn => btn.classList.remove('aquaponic__btn-active'));

    // give this class active
    evt.target.classList.add('aquaponic__btn-active');

    //set id of input field
    const input = document.getElementsByClassName('aquaponic__input')[0];
    if (input) input.setAttribute('id', evt.target.innerHTML);
    if (input) document.getElementsByClassName('aquaponic__input')[0].focus(); //send cursor to input when btn clicked

    this.props.setSelectedFilter(evt.target.innerHTML);

    // hide the non-selected data visualisation (lines)
    const line1 = document.getElementsByClassName('aquaponic__line1')[0];
    const line2 = document.getElementsByClassName('aquaponic__line2')[0];
    const line3 = document.getElementsByClassName('aquaponic__line3')[0];
    const line4 = document.getElementsByClassName('aquaponic__line4')[0];
    const dots1 = document.getElementsByClassName('circle-ph');
    const dots2 = document.getElementsByClassName('circle-amm');
    const dots3 = document.getElementsByClassName('circle-ite');
    const dots4 = document.getElementsByClassName('circle-ate');

    // hide the non-selected graph information
    switch (evt.target.innerHTML) {
      case 'pH':
        line2.style.opacity = 0;
        line3.style.opacity = 0;
        line4.style.opacity = 0;
        for (let dot of dots2) dot.style.opacity = 0;
        for (let dot of dots3) dot.style.opacity = 0;
        for (let dot of dots4) dot.style.opacity = 0;
        break;
      case 'ammonia':
        line1.style.opacity = 0;
        line3.style.opacity = 0;
        line4.style.opacity = 0;
        for (let dot of dots1) dot.style.opacity = 0;
        for (let dot of dots3) dot.style.opacity = 0;
        for (let dot of dots4) dot.style.opacity = 0;
        break;
      case 'nitrite':
        line1.style.opacity = 0;
        line2.style.opacity = 0;
        line4.style.opacity = 0;
        for (let dot of dots1) dot.style.opacity = 0;
        for (let dot of dots2) dot.style.opacity = 0;
        for (let dot of dots4) dot.style.opacity = 0;
        break;
      case 'nitrate':
        line2.style.opacity = 0;
        line3.style.opacity = 0;
        line1.style.opacity = 0;
        for (let dot of dots2) dot.style.opacity = 0;
        for (let dot of dots3) dot.style.opacity = 0;
        for (let dot of dots1) dot.style.opacity = 0;
        break;
        // case all
      default: //view all
        break;
    }
  };

  handleOnInputChange = evt => {
    const value = evt.target.value;
    
    if (this.state.selected === 'pH') { 
      const regex = /^(1[0-3](\.\d{0,1})?)$|^\d?(\.\d{0,1})?$|^14$/; // 1 -> 14 with opt .0
      if (!value || value.match(regex)) {
          console.log(`ph: ${value}`);
          this.setState(() => ({ input: value, error: '' }));
        } else {
          this.setState(() => ({ error: `${this.state.selected} range is 1 -> 14` }));
        }
        
    } else if (this.state.selected === 'nitrate') { //0 -> 200 with opt .0
      const regex = /^1?\d?\d?(\.\d{0,1})?$|^200$/;
      if (!value || value.match(regex)) {
        console.log(`nitrate: ${value}`);
        this.setState(() => ({ input: value, error: '' }));
      } else {
        this.setState(() => ({ error: 'nitrate should be in the range 0 -> 120 ppm' }));
      }

    } else {
      const regex = /^10(\.0?)?$|^\d(\.\d{0,1})?$/; //ammonia or nitrite, 0 -> 10 
      if (!value || value.match(regex)) {
        console.log(`amm/nitrite: ${value}`);
        this.setState(() => ({ input: value, error: '' }));
      } else {
        this.setState(() => ({ error: `${this.state.selected} should be in the range 0 -> 10 ppm` }));
      }
    }
  }
  
  onSubmit = evt => {
    evt.preventDefault();
    const input = document.querySelector('.aquaponic__input');
    let val = input.value;
    if (val.endsWith('.')) val = val += '0'; // if entry ends in decimal, concat a 0
    const obj = {value: val, name: this.state.selected, date: new Date().toString()};
    this.props.startAddEntry(obj)
    this.setState(() => ({ input: '', error: '' }));
  };

  render() {
    return (
      <div className="aquaponic">
        <section className="content-container aquaponic__flex-container">
          <div className="aquaponic__btn-bar">
            <button className="aquaponic__btn aquaponic__btn--ph" onClick={this.handleOnClick}>
              pH
            </button>
            <button className="aquaponic__btn aquaponic__btn--amm" onClick={this.handleOnClick}>
              ammonia
            </button>
            <button className="aquaponic__btn aquaponic__btn--ite" onClick={this.handleOnClick}>
              nitrite
            </button>
            <button className="aquaponic__btn aquaponic__btn--ate" onClick={this.handleOnClick}>
              nitrate
            </button>
            <button
              className="aquaponic__btn aquaponic__btn-active"
              onClick={this.handleOnClick}
            >
              view all
            </button>
            <div className="content-container flex-container">
              <form
                className="aquaponic__form"
                onSubmit={this.onSubmit}
                autoComplete="off"
              >
                {this.state.selected !== 'view all' && (
                  <p> How much <span>{this.state.selected}</span> today?</p>
                )}

                {this.state.selected !== 'view all' && (
                  <input
                    autoFocus
                    className="aquaponic__input"
                    type="text"
                    id="pH"
                    value={this.state.input}
                    onChange={this.handleOnInputChange}
                  />
                )}
                {this.state.error && (
                  <p className="aquaponic__error-msg">{this.state.error}</p>
                )}
              </form>
            </div>
          </div>
          {console.log(window.innerWidth)          }
          {window.innerWidth < 850 ? <D3LineGraph windowWidth={window.innerWidth }/>
          : <D3LineGraph />
          }
        </section>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  setSelectedFilter: wqm => dispatch(setSelectedFilter(wqm)),
  startAddEntry: entry => dispatch(startAddEntry(entry)),
});

export default connect(undefined,mapDispatchToProps)(Aquaponic);