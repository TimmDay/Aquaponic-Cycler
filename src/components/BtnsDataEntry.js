import React from 'react';
import { connect } from 'react-redux';
import { setSelectedFilter } from '../actions/aquaponic';
import InputDataEntry from './InputDataEntry';

class BtnsDataEntry extends React.Component {
   constructor(props) {
       super(props);
       this.state = {
        selected: 'view all',
       }
   }

   handleOnClick = evt => {
    // update text of form span
    this.setState({ selected: evt.target.innerHTML });

    // remove active btn class from others
    const els = document.querySelectorAll('.aquaponic__btn');
    els.forEach(btn => btn.classList.remove('aquaponic__btn-active'));

    // give this class active
    evt.target.classList.add('aquaponic__btn-active');

    //set id of input field
    const input = document.getElementsByClassName('aquaponic__input')[0];
    if (input) input.setAttribute('id', evt.target.innerHTML);
    if (input) document.getElementsByClassName('aquaponic__input')[0].focus(); //send cursor to input when btn clicked

    this.props.setSelectedFilter(evt.target.innerHTML);

    // hide the non-selected data visualisation (lines)
    const line1 = document.getElementsByClassName('aquaponic__line1')[0];
    const line2 = document.getElementsByClassName('aquaponic__line2')[0];
    const line3 = document.getElementsByClassName('aquaponic__line3')[0];
    const line4 = document.getElementsByClassName('aquaponic__line4')[0];
    const dots1 = document.getElementsByClassName('circle-ph');
    const dots2 = document.getElementsByClassName('circle-amm');
    const dots3 = document.getElementsByClassName('circle-ite');
    const dots4 = document.getElementsByClassName('circle-ate');

    // hide the non-selected graph information
    switch (evt.target.innerHTML) {
      case 'pH':
        line2.style.opacity = 0;
        line3.style.opacity = 0;
        line4.style.opacity = 0;
        for (let dot of dots2) dot.style.opacity = 0;
        for (let dot of dots3) dot.style.opacity = 0;
        for (let dot of dots4) dot.style.opacity = 0;
        break;
      case 'ammonia':
        line1.style.opacity = 0;
        line3.style.opacity = 0;
        line4.style.opacity = 0;
        for (let dot of dots1) dot.style.opacity = 0;
        for (let dot of dots3) dot.style.opacity = 0;
        for (let dot of dots4) dot.style.opacity = 0;
        break;
      case 'nitrite':
        line1.style.opacity = 0;
        line2.style.opacity = 0;
        line4.style.opacity = 0;
        for (let dot of dots1) dot.style.opacity = 0;
        for (let dot of dots2) dot.style.opacity = 0;
        for (let dot of dots4) dot.style.opacity = 0;
        break;
      case 'nitrate':
        line2.style.opacity = 0;
        line3.style.opacity = 0;
        line1.style.opacity = 0;
        for (let dot of dots2) dot.style.opacity = 0;
        for (let dot of dots3) dot.style.opacity = 0;
        for (let dot of dots1) dot.style.opacity = 0;
        break;
      default: //view all
        break;
    }
  };

   render () {
       return (
        <div className="aquaponic__btn-bar">
          <button className="aquaponic__btn aquaponic__btn--ph" onClick={this.handleOnClick}>pH</button>
          <button className="aquaponic__btn aquaponic__btn--amm" onClick={this.handleOnClick}>ammonia</button>
          <button className="aquaponic__btn aquaponic__btn--ite" onClick={this.handleOnClick}>nitrite</button>
          <button className="aquaponic__btn aquaponic__btn--ate" onClick={this.handleOnClick}>nitrate</button>
          <button className="aquaponic__btn aquaponic__btn-active" onClick={this.handleOnClick}>view all</button>
          
          < InputDataEntry selected={this.state.selected} />
          
        </div>
       )
   }
}

const mapDispatchToProps = (dispatch) => ({
  setSelectedFilter: wqm => dispatch(setSelectedFilter(wqm))
});

export default connect(undefined, mapDispatchToProps)(BtnsDataEntry);