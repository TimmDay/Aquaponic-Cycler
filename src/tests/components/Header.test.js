import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

test('should render Header correctly', () => {
  const wrapper = shallow(<Header startLogout={() => { }} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('h1').text()).toBe('Aquaponic Cycler')
});

test('should call startLogout on button click', () => {
  const startLogout = jest.fn();
  const wrapper = shallow(<Header startLogout={startLogout} />);
  wrapper.find('button').simulate('click');
  expect(startLogout).toHaveBeenCalled();
});
