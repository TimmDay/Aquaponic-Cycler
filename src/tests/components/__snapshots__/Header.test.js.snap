// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/aquaponic"
      >
        <h1>
          Aquaponic Cycler
        </h1>
        <p
          className="header_tagline"
        >
          Tracking your bacterial food producers
        </p>
      </Link>
      <button
        className="header__link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
